code/parse/sexp.h

```diff
@@ -1111,6 +1119,9 @@ struct sexp_cached_data
 	int numeric_literal = 0;				// i.e. a number
 	int ship_registry_index = -1;			// because ship status is pretty common
 	void *pointer = nullptr;				// could be an IFF, a wing, a goal, or other unchanging reference
+	// jg18 - used to ensure sexp_container_CTEXT() is idempotent
+	char sexp_node_text[TOKEN_LENGTH] = "";
```

Comment:
There is already a `Sexp_nodes[].text` field.  As with the other review comment, I think I need to know more about how the design of containers is supposed to work.

Reply:
The reason is that a given instance of "Replace Container" should be evaluated only once, then the computed value should be used after that. I'll work on creating an examplw.

With a node that uses "Replace Variable", is the latest value retrieved every time the node is processed with `CTEXT()`? My impression from the code is "yes".

Reply:
Err, after noticing that text replacement in a message is applied anew each time the message is sent, maybe we *do* want Replace Container on a node to be evaluated more than once?

The tricky aspect is that some list container modifiers, such as `Remove_First`, have side effects, and we wanted to keep `CTEXT()` idempotent.

One (partial?) solution is to perform the side effect only when containers should be allowed to be modified, such as in-mission (see `are_containers_modifiable()` in sexp_container.cpp).

Reply:
Another option is to disallow in "Replace Container" the list modifiers that have side effects (`Remove_First`, `Remove_Last`, `Remove_Random`) and support their functionality through new SEXPs instead. These modifiers could still be supported in text replacement.

But what effect would that change have on the feature's usefulness for FREDders? @Karajorma any thoughts?

Reply:
I renamed the field to `container_CTEXT_result` and removed the `bool` field.

The problem this solves is that `sexp_container_CTEXT()` dynamically generates its result and thus we require space to store it. In the cases of list modifiers that remove the retrieved value (e.g., `Remove_First`), the original value won't even exist anymore by the time `sexp_container_CTEXT()` returns.

Reply:
Oh wait, now I get your concern. Yeah, I think you're right, we should be updating the text in `tree_nodes` instead and then allowing FRED to update `Sexp_nodes` when it saves its view of the SEXP tree.

If that's the case, then if the FREDder makes changes in the Add/Modify Container dialog and clicks OK om that dialog *but* then presses Cancel in the Events Editor, should none of the container changes be applied to `Sexp_containers`? Does modifying variables work that way?

If so, then the current Add/Modify Container dialog behavior is incorrect, because that dialog updates `Sexp_containers` after you click OK in it, regardless of what you click in the Events Editor. I'll create a separate PR to fix that.